src/test/java/org/littleshoot/proxy/DirectRequestTest.java

```diff
@@ -0,0 +1,108 @@
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.littleshoot.proxy.impl.DefaultHttpProxyServer;
+import org.littleshoot.proxy.test.SocketClientUtil;
+
+/**
+ * This class tests direct requests to the proxy server, which causes endless
+ * loops (#205).
+ */
+public class DirectRequestTest {
+    private HttpProxyServer proxyServer;
+
+    private Socket socket;
+
+    @Before
+    public void setUp() throws Exception {
+        socket = null;
+        proxyServer = null;
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        try {
+            if (proxyServer != null) {
+                proxyServer.abort();
+            }
+        } finally {
+            if (socket != null) {
+                socket.close();
+            }
+        }
+    }
+
+    @Test(timeout = 5000)
+    public void testDirectRequestAnsweredBadRequest() throws Exception {
+
+        HttpFiltersSource filtersSource = new HttpFiltersSourceAdapter() {
+            @Override
+            public HttpFilters filterRequest(HttpRequest originalRequest) {
+                return new HttpFiltersAdapter(originalRequest);
+            }
+        };
+
+        proxyServer = DefaultHttpProxyServer.bootstrap()//
+                .withPort(0)//
+                .withFiltersSource(filtersSource)//
+                .start();
+
+        socket = SocketClientUtil.getSocketToProxyServer(proxyServer);
+
+        int proxyPort = proxyServer.getListenAddress().getPort();
+        String successfulGet = "GET http://localhost:" + proxyPort + "/directToProxy HTTP/1.1\n" + "\r\n";
+        SocketClientUtil.writeStringToSocket(successfulGet, socket);
+
+        Thread.sleep(750);
+
+        String response = SocketClientUtil.readStringFromSocket(socket);
+
+        assertThat("Expected to receive an HTTP 400 from the server", response, startsWith("HTTP/1.1 400 "));
+    }
+
+    @Test(timeout = 5000)
+    public void testDirectRequestAnswerUnmodified() throws Exception {
+
+        HttpFiltersSource filtersSource = new HttpFiltersSourceAdapter() {
+            @Override
+            public HttpFilters filterRequest(HttpRequest originalRequest) {
+                return new HttpFiltersAdapter(originalRequest) {
+                    @Override
+                    public HttpResponse clientToProxyRequest(HttpObject httpObject) {
+                        return new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.valueOf(403));
+                    }
+                };
+            }
+        };
+
+        proxyServer = DefaultHttpProxyServer.bootstrap()//
+                .withPort(0)//
+                .withFiltersSource(filtersSource)//
+                .start();
+
+        socket = SocketClientUtil.getSocketToProxyServer(proxyServer);
+
+        int proxyPort = proxyServer.getListenAddress().getPort();
+        String successfulGet = "GET http://localhost:" + proxyPort + "/directToProxy HTTP/1.1\n" + "\r\n";
```

Comment:
Yes, it returns 400 with 127.0.0.1 too. I've always done this in the 
browser. Thank you, should add a test for this, just copied from 
KeepAliveTest :-)

I've tested only the external IP4 adress. Getting the external host name 
of the proxy causes a blocking behavior, but no endless loop. I will 
investigate for this and for IP6 adresses too.

Am 19.07.2015 um 21:58 schrieb Jason Hoetger:
> In src/test/java/org/littleshoot/proxy/DirectRequestTest.java
> <https://github.com/adamfisk/LittleProxy/pull/220#discussion_r34961493>:
>
>> +                    public HttpResponse clientToProxyRequest(HttpObject httpObject) {
>> +                        return new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.valueOf(403));
>> +                    }
>> +                };
>> +            }
>> +        };
>> +
>> +        proxyServer = DefaultHttpProxyServer.bootstrap()//
>> +                .withPort(0)//
>> +                .withFiltersSource(filtersSource)//
>> +                .start();
>> +
>> +        socket = SocketClientUtil.getSocketToProxyServer(proxyServer);
>> +
>> +        int proxyPort = proxyServer.getListenAddress().getPort();
>> +        String successfulGet = "GET http://localhost:" + proxyPort + "/directToProxy HTTP/1.1\n" + "\r\n";
>
> Does this PR also return 400s when the client makes a request when the
> hostname is not localhost? For example, http://127.0.0.1, or
> http://some-machine-name, assuming "some-machine-name" resolves to an
> address the proxy is listening on.
>
> (Not saying it has to, just wondering if it does.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/adamfisk/LittleProxy/pull/220/files#r34961493>.
>